pychiquito/dsl.py

```diff
@@ -2,83 +2,159 @@
+        print(f"first step id: {step_type.step_type.id}")
+
+    def pragma_last_step(self: Circuit, step_type: StepType) -> None:
+        assert self.mode == CircuitMode.SETUP
+        self.ast.last_step = step_type.step_type.id
+        print(f"last step id: {step_type.step_type.id}")
+
+    def pragma_num_steps(self: Circuit, num_steps: int) -> None:
+        assert self.mode == CircuitMode.SETUP
+        self.ast.num_steps = num_steps
+
+    def pragma_disable_q_enable(self: Circuit) -> None:
+        assert self.mode == CircuitMode.SETUP
+        self.ast.q_enable = False
+
+    def gen_witness(self: Circuit, args: Any) -> TraceWitness:
+        self.mode = CircuitMode.Trace
+        self.trace_context = TraceContext()
+        self.ast.set_trace(self.trace)
+        self.trace(args)
+        return self.trace_context.witness
+
+    def add(self: Circuit, step_type: StepType, args: Any):
+        print(self)
+        print(step_type)
+        print(args)
+        assert self.mode == CircuitMode.Trace
+        self.trace_context.add(self, step_type, args)
+
+    def print_ast(self: Circuit):
+        print("Print ASTCircuit using custom __str__ method in python:")
+        print(self.ast)
+
+    def get_ast_json(self: Circuit, print_json=False) -> str:
+        ast_json: str = json.dumps(self.ast, cls=CustomEncoder, indent=4)
+        if print_json:
+            print("Print ASTCircuit using __json__ method in python:")
+            print(ast_json)
+        return ast_json
+
+    def print_witness(self: Circuit):
+        print("Print TraceWitness using custom __str__ method in python:")
+        print(self.trace_context.witness)
+
+    def get_witness_json(self: Circuit, print_json=False) -> str:
+        witness_json: str = json.dumps(
+            self.trace_context.witness, cls=CustomEncoder, indent=4
         )
-        return step_type_context
-
-    def step_type_def(self: StepTypeContext) -> StepTypeContext:
-        self.circuit.add_step_type_def()
-
-    def trace(
-        self: CircuitContext, trace_def: Callable[[TraceContext, Any], None]
-    ):  # TraceArgs are Any.
-        self.circuit.set_trace(trace_def)
-
-    def fixed_gen(
-        self: CircuitContext, fixed_gen_def: Callable[[FixedGenContext], None]
-    ):
-        self.circuit.set_fixed_gen(fixed_gen_def)
-
-    def pragma_first_step(
-        self: CircuitContext, step_type_context: StepTypeContext
-    ) -> None:
-        self.circuit.first_step = step_type_context.step_type.id
-        print(f"first step id: {step_type_context.step_type.id}")
-
-    def pragma_last_step(
-        self: CircuitContext, step_type_context: StepTypeContext
-    ) -> None:
-        self.circuit.last_step = step_type_context.step_type.id
-        print(f"last step id: {step_type_context.step_type.id}")
-
-    def pragma_num_steps(self: CircuitContext, num_steps: int) -> None:
-        self.circuit.num_steps = num_steps
+        if print_json:
```

Comment:
@qwang98 agains this is kinf of user interface or debugging, it does not belong here

Reply:
Fixed.

Reply:
Fixed!